omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,899 @@
+            } else {
+                omp_unset_lock(&read_lock);
+                return std::make_pair(nullptr, 0);
+            }
+        }
+    }
+};
+
+
+/**
+ * Performs an unordered breadth-first search,
+ * thereby building a rooted level structure.
+ */
+template <typename IndexType>
+void ubfs(std::shared_ptr<const OmpExecutor> exec, const size_type num_vertices,
+          const IndexType *const row_ptrs, const IndexType *const col_idxs,
+          IndexType *const
+              levels,  // Must be inf/max in all nodes connected to source
+          const IndexType start, const IndexType max_degree)
+{
+    const auto max_threads = omp_get_max_threads();
+
+    // This is an upper bound for how many nodes may be in the work set
+    // at maximum at any time.
+    const auto work_bound = (size_type)max_threads * num_vertices;
+    UbfsLinearQueue<IndexType> q(exec, work_bound);
+
+    // Start from the given start node.
+    q.enqueue_chunk(&start, 1);
+    levels[start] = 0;
+
+    // Used for synchronization of (pseudo)remaining work.
+    int threads_working = 0;
+
+#pragma omp parallel firstprivate(max_degree)
+    {
+        // This is an upper bound, as '.deque_chunk...' is upper bounded by
+        // chunk_bound.
+        gko::vector<IndexType> local_to_insert(max_degree * chunk_bound, exec);
+        auto local_to_insert_size = 0;
+
+        while (true) {
+            // Get some nodes from the locked q.
+            const auto pair = q.dequeue_chunk(&threads_working);
+            const auto chunk = pair.first;
+            const auto chunk_size = pair.second;
+
+            // If there are nodes left to process.
+            // If there are no nodes returned that also means there are no nodes
+            // still in flight.
+            if (chunk_size > 0) {
+                // For each node in the chunk.
+                for (auto chunk_i = 0; chunk_i < chunk_size; ++chunk_i) {
+                    const auto node = chunk[chunk_i];
+
+                    IndexType level;
+#pragma omp atomic read
+                    level = levels[node];
+
+                    const auto new_neighbour_level = level + 1;
+
+                    // Go through neighbours.
+                    const auto row_start = row_ptrs[node];
+                    const auto row_end = row_ptrs[node + 1];
+
+                    for (auto neighbour_i = row_start; neighbour_i < row_end;
+                         ++neighbour_i) {
+                        const auto neighbour = col_idxs[neighbour_i];
+
+                        IndexType old_neighbour_level;
+                        do {
+#pragma omp atomic read
+                            old_neighbour_level = levels[neighbour];
+
+                            const auto is_new_smaller =
+                                new_neighbour_level < old_neighbour_level;
+
+                            if (!is_new_smaller) {
+                                goto continue_outer;
```

Comment:
That's exactly the xkcd I thought of while writing this ;-)
On the other hand, in combination with the compare-exchange, I think this code is already sufficiently complex. Also, can we be sure this compiles to similarly optimal code?

Reply:
Thinking about it again, I would prefer if you could wrap this into an immediately evaluated lambda expression (you probably don't need a return type, right?)
I would expect the codegen to be exactly equivalent, a really simple example seems to confirm that (https://godbolt.org/z/v58Wjo)
Since you already correctly noted that this function is rather complex, do you see a good way to extract some parts of it into a separate function?

Reply:
Alright then, I'll be right at it.